Core Data: Update RefundStore to move write operations to background

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #14090 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR updated `RefundStore` following the updates of the `CoreDataManager` and optimize its fetch requests. Specifically:
- Removed deprecated access to `sharedDerivedStorage` and switch to using `performAndSave`.
- Updated `retrieveRefunds` to delete stale refunds as part of the upsertion to ensure the deletion is done in the background.
- Updated `resetStoredRefunds` and `deleteStoredRefund` to ensure the deletion is done in the background.
- Optimized fetch requests in `upsertStoredRefunds`.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Log in to a test store with existing products.
- Create an order with more than 1 product or open an existing one.
- Select Issue Refund and refund 1 or more item in the order.
- Confirm that refund creation succeeds and the refunded items are then listed in the order details.
- Log out of the store and log back in.
- Select the the order with the refunded items and confirm that the refund details are still correct.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested on simulator iPhone 16 Pro iOS 18.1 and confirmed that creating and fetching refunds work as expected.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
TODO: @itsmeichigo to update the release notes before merging

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

itsmeichigo committed:
Remove deprecated access to sharedDerivedStorage in RefundStore

itsmeichigo committed:
Update deleteStoredRefund to delete item in background

itsmeichigo committed:
Update resetStoredRefunds to remove items in background

itsmeichigo committed:
Optimize fetching requests for RefundStore

itsmeichigo committed:
Update retrieveRefunds to delete items in background

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Remove redundant space

itsmeichigo committed:
Add tests for new storage type extension methods

itsmeichigo committed:
Use stored refund items and shipping lines other then fetching

itsmeichigo committed:
Merge branch 'trunk' into task/14090-refund-store-update

selanthiraiyan commented on Yosemite/Yosemite/Stores/RefundStore.swift:
```diff
@@ -177,9 +208,13 @@ private extension RefundStore {
         for readOnlyRefund in readOnlyRefunds {
-            let storageRefund = storage.loadRefund(siteID: readOnlyRefund.siteID, orderID: readOnlyRefund.orderID, refundID: readOnlyRefund.refundID) ??
+            let storageRefund = storedRefunds.first(where: { $0.refundID == readOnlyRefund.refundID }) ??
                 storage.insertNewObject(ofType: Storage.Refund.self)
```
Super nit: Indentation needs to be updated.

itsmeichigo replied:
Done in 8d7ccc274c3196c8547027ae6c0a27f17ac543b8 (I hope this is what you mean)

selanthiraiyan reviewed (approved):
I tested that order refund works as described. The code LGTM. 

Thanks for optimizing the fetch request logic. 👏 

---

The PR is good to merge after addressing the following points. :shipit: 

1. Not created by this PR: There are two unused constants in `func handleOrderItemRefunds`. Kindly remove that.
2. The `Testing information` section is empty in PR description. Kindly consider updating it by adding the flows that you tested.

itsmeichigo committed:
Remove unused variables

itsmeichigo committed:
Update release notes

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/228">20.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented:
Thanks @selanthiraiyan for the reviews! I updated the PR description and addressed your comments, I hope everything looks good now. I'll merge the PR later today unless there are any other comments.
